packages/styled/types/tests.tsx

```diff
@@ -217,3 +217,69 @@ const Input5 = styled.input`
   // $ExpectError
   ;<StyledCompWithoutAs as={Section} />
 }
+
+{
+  // Props forwarding through StyledOptions and FilteringStyledOptions
+
+  const fc: React.FC<{ foo: string }> = props => <div {...props} />
+
+  styled(fc, { shouldForwardProp: (prop: 'foo') => true })({})
+
+  styled(fc, { shouldForwardProp: (prop: string) => true })({})
+
+  // $ExpectError
+  styled(fc, { shouldForwardProp: (prop: 'bar') => true })({})
+
+  const shouldForwardProp1 = (prop: 'foo') => true
+  styled(fc, { shouldForwardProp: shouldForwardProp1 })({})
+
+  const shouldForwardProp2: StyledOptions['shouldForwardProp'] = (
+    prop: 'unknown'
+  ) => true
+  styled(fc, { shouldForwardProp: shouldForwardProp2 })({})
+
+  const shouldForwardProp3 = (prop: 'unknown') => true
+  // $ExpectError
+  styled(fc, { shouldForwardProp: shouldForwardProp3 })({})
+
+  // $ExpectError
+  const shouldForwardProp4: StyledOptions<{
+    foo: string
+  }>['shouldForwardProp'] = (prop: 'unknown') => true
+
+  const shouldForwardProp5 = (prop: 'foo'): prop is 'foo' => true
+  styled(fc, { shouldForwardProp: shouldForwardProp5 })({})
+
+  const shouldForwardProp6: FilteringStyledOptions['shouldForwardProp'] = (
+    prop: 'foo'
+  ): prop is 'foo' => true
+
+  const shouldForwardProp7: FilteringStyledOptions<{
+    foo: string
+  }>['shouldForwardProp'] = (prop: 'foo'): prop is 'foo' => true
+
+  // $ExpectError
+  const shouldForwardProp8: FilteringStyledOptions<{
+    foo: string
+  }>['shouldForwardProp'] = (prop: 'unknown'): prop is 'unknown' => true
+
+  const shouldForwardProp9: FilteringStyledOptions<
+    { foo: string; bar: string },
+    'foo'
+  >['shouldForwardProp'] = (prop: 'foo'): prop is 'foo' => true
```

Comment:
Perhaps interestingly this is types as:
```
const shouldForwardProp9: ((propName: string | number | symbol) => propName is "foo") | undefined
```

Could we add comment above `FilteringStyledOptions` and `StyledOptions` mentioning what are their responsibilities? I am somewhat confused about that. I understand that for filtering variant we accept a type predicate whereas for the other one we accept a simpler function that returns a boolean **but** that is not the only difference between them and I feel like this should be clarified.

Reply:
Smth like this?
```ts
/**
 * Same as StyledOptions but shouldForwardProp must be a type guard (https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates).
 * Practical meaning: you can declare and reuse atomic `shouldForwardProp` filters that are strictly bound with some `ForwardedProps` type.
 */
```

Reply:
ye, i guess that should suffice